Implicit Caller Location section.

anp opened the pull request:
Documenting the overall feature here as the current implementation is spread across several PRs and has deviated a good bit from the RFC.

recommended in https://github.com/rust-lang/rust/issues/47809#issuecomment-572497766

eddyb commented:
Probably an easy way to explain it is to show the definition and what it codegens as, both in regular Rust code (even if nothing before codegen, not even MIR, encodes the caller location passing).

eddyb commented on src/codegen/implicit-caller-location.md:
```diff
@@ -0,0 +1,154 @@
+The `#[track_caller]` attribute is checked alongside other codegen attrs to ensure the function:
+
+* has `"Rust"` ABI (as opposed to `"C"`, etc...)
+* is not in an extern block
```
We probably use "extern block" elsewhere, but I wish we could say "is not a foreign import". or "FFI import'

Eh, at the very least, maybe write `extern {...}` instead of "extern".

anp commented:
cc @nikomatsakis you had specifically asked for this to be able to learn more about the feature. Would love to hear if it achieves that goal!

yaahc commented on src/codegen/implicit-caller-location.md:
```diff
@@ -0,0 +1,278 @@
+rather than walking up the stack when the intrinsic is called. That additional argument can be
+returned wherever the caller location is queried.
+
+The argument we append is of type `&'static core::panic::Location<'staic>`. A reference was chosen
```
```suggestion
The argument we append is of type `&'static core::panic::Location<'static>`. A reference was chosen
```

yaahc commented on src/codegen/implicit-caller-location.md:
```diff
@@ -0,0 +1,278 @@
+[`caller_location`][intrinsic] compiler intrinsic implemented by rustc. This intrinsic is easiest 
+explained in terms of how it works in a `const` context.
+
+## Caller Location in `const`
```
This is information dense and I'm not sure its actually clearer than the next section, to me adding an extra arg where we forward information down every time its a `#[track_caller]` is much more intuitive

anp replied:
I think I've addressed this, let me know if you still find it dense!

yaahc commented on src/codegen/implicit-caller-location.md:
```diff
@@ -0,0 +1,278 @@
+trait method. When applied to a default trait method implementation, the attribute takes effect on
+that implementation *and* any overrides. It is valid to apply the attribute to a regular
+implementation of a trait method, regardless of whether the defining trait does. It is a no-op to
+apply the attribute to trait methods with the attribute in the trait definition, but a valid one.
```
the last two sentences here were a little difficult to follow

nikomatsakis commented on src/codegen/implicit-caller-location.md:
```diff
@@ -0,0 +1,278 @@
+location, so functions which intentionally invoked `panic!` couldn't provide their own location, 
+hiding the actual source of error.
+
+Internally, `panic!()` now calls [`core::panid::Location::caller()`][wrapper] to find out where it 
```
```suggestion
Internally, `panic!()` now calls [`core::panic::Location::caller()`][wrapper] to find out where it 
```

nikomatsakis commented on src/codegen/implicit-caller-location.md:
```diff
@@ -0,0 +1,278 @@
+If the calling function is not tracked, `get_caller_location` allocates a `Location` static from
+the current `Span` and returns a reference to that.
+
+By chaining together the `caller_location` fields of multiple `FunctionCx`s as we grow the bottom of
```
Nit: the term "chaining together" is confusing to me. Are we building a linked list in memory?

nikomatsakis replied:
Oh, I guess we are. OK, great.

eddyb replied:
Not AFAIK, it's more of a conceptual chaining, every `#[track_caller]` function using their location parameter, to which a `#[track_caller]` caller would've passed in their own location parameter etc.

anp replied:
`We achieve the same behavior as a loop starting from the bottom by passing a single `&Location` value through the `caller_location` fields of multiple `FunctionCx`s as we grow the stack downward."

?

anp replied:
Actually just going to commit that and push, it's the last bit of feedback.

anp commented on src/codegen/implicit-caller-location.md:
```diff
@@ -0,0 +1,278 @@
+reach the first function call in the stack which does *not* have the attribute. This walk is in
+[`InterpCx::find_closest_untracked_caller_location()`][const-find-closest].
+
+Starting at the bottom, we iterate up over stack [`Frame`][const-frame]s in the [InterpCx::stack],
```
Good catch! The anchor reference needs to be updated too, which reflows the paragraph, which...

anp commented:
OK! Thank you for the reviews and the kind words <3. I think I've addressed all the feedback.

Once this is re-approved, it's gated on https://github.com/rust-lang/rustc-dev-guide/pull/641 and https://github.com/rust-lang/rustc-dev-guide/pull/646 to get CI green. I'll add a link from the code to this once it lands?

mark-i-m commented:
@anp Could you rebase on top of the most recent master? It has a bunch of link fixes.

anp commented:
Actually just finished doing that, those breakages were from my own links 😅 -- fixed up.
